core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -836,8 +647,12 @@ class KafkaController(val config: KafkaConfig,
     // New reassignments may have been submitted through Zookeeper while the controller was failing over
     val zkPartitionsResumed = processZkPartitionReassignment()
     // We may also have some API-based reassignments that need to be restarted
-    maybeResumeReassignments { (tp, _) =>
-      !zkPartitionsResumed.contains(tp)
+    try {
+      reassignmentsManager.maybeResumeReassignments { (tp, _) =>
+        !zkPartitionsResumed.contains(tp)
+      }
+    } catch {
+      case e: IllegalStateException => handleIllegalState(e)
```

Comment:
I did this because `maybeResumeReassignments` calls `onPartitionReassignment` which calls `sendRequestsToBrokers` in phase B or in phase A's `updateLeaderEpochAndSendRequest` call.

I wanted to ensure that any errors there are caught, otherwise this patch would again change the behavior in it being propagated. If we don't catch this, it is again a behavioral change.

Is there concern that other methods may raise an IllegalStateException?
Perhaps we can re-throw the `sendRequestsToBrokers`' exception to something else and catch that only?

Reply:
Yeah, my concern was that something else might raise IllegalStateException. I am actually a bit tempted to get rid of `handleIllegalState` altogether. It just seems so arbitrary. Looking through the code, I cannot see the specific case we're trying to protect. The call to `sendRequestsToBrokers` just builds the request and puts it in a queue. Perhaps I'm missing something?

Reply:
The `handleIllegalState` stuff was added by Flavio to fix a bug that had been reported. We kept that through the Controller refactoring. A few related PRs:
* https://github.com/apache/kafka/pull/102
* https://github.com/apache/kafka/pull/212
* https://github.com/apache/kafka/pull/2935

Reply:
Thanks @ijuma. That helps. It looks like it was specifically trying to protect the validation we do in `newBatch()`, but the root cause was evidently unknown. And neither have I heard of any recurrence of the issue.  So my first inclination after seeing this is to also get rid of `newBatch` along with the logic to resign the controller (which seems like massive overkill). Will look more carefully tomorrow if I get a chance.